core/templates/dev/head/domain/objects/UnitsObjectFactory.js

```diff
@@ -0,0 +1,205 @@
+
+/**
+ * @fileoverview Factory for creating instances of Units domain objects.
+ */
+
+oppia.factory('UnitsObjectFactory', ['CURRENCY_UNITS',
+  function(CURRENCY_UNITS) {
+    var Units = function(unitsList) {
+      this.units = unitsList;
+    };
+
+    var isunit = function(unit) {
+      return !('/*() '.includes(unit));
+    };
+
+    Units.stringToLexical = function(units) {
+      units += '#';
+      var unitList = [];
+      var unit = '';
+      for (var i = 0; i < units.length; i++) {
+        if ('*/()# '.includes(units[i]) && unit !== 'per') {
+          if (unit.length > 0) {
+            if ((unitList.length > 0) && isunit(unitList.slice(-1).pop())) {
+              unitList.push('*');
+            }
+            unitList.push(unit);
+            unit = '';
+          }
+          if (!('# '.includes(units[i]))) {
+            unitList.push(units[i]);
+          }
+        } else if (units[i] === ' ' && unit === 'per') {
+          unitList.push('/');
+          unit = '';
+        } else {
+          unit += units[i];
+        }
+      }
+      return unitList;
+    };
+
+    var unitWithMultiplier = function(unitList) {
+      var multiplier = 1;
+      var unitsWithMultiplier = [];
+      var parenthesisStack = [];
+
+      for (var ind = 0; ind < unitList.length; ind++) {
+        if (unitList[ind] === '/') {
+          multiplier = -multiplier;
+        } else if (unitList[ind] === '(') {
+          if (unitList[ind - 1] === '/') {
+            // If previous element was division then we need to inverse
+            // multiplier when we find its corresponsing closing parenthesis.
+            // Second element of pushed element is used for this purpose.
+            parenthesisStack.push(['(', -1]);
+          } else {
+            // If previous element was not division then we don't need to
+            // invert the multiplier.
+            parenthesisStack.push(['(', 1]);
+          }
+        } else if (unitList[ind] === ')') {
+          var elem = parenthesisStack.pop();
+          multiplier = parseInt(elem[1]) * multiplier;
+        } else if (isunit(unitList[ind])) {
+          unitsWithMultiplier.push([unitList[ind], multiplier]);
+          // If previous element was division then we need to invert
+          // multiplier.
+          if (unitList[ind - 1] === '/') {
+            multiplier = -multiplier;
+          }
+        }
+      }
+      return unitsWithMultiplier;
+    };
+
+    var convertUnitDictToList = function(unitDict) {
+      var unitList = [];
+      for (var key in unitDict) {
+        unitList.push({unit: key, exponent: unitDict[key]});
+      }
+      return unitList;
+    };
+
+    var unitToList = function(unitsWithMultiplier) {
+      var unitDict = {};
+      for (var i = 0; i < unitsWithMultiplier.length; i++) {
+        var unit = unitsWithMultiplier[i][0];
+        var multiplier = unitsWithMultiplier[i][1];
+        var ind = unit.indexOf('^');
+        if (ind > -1) {
```

Comment:
Move declaration of s and power before the if statement.

Reply:
Done